### PR TITLE
Add ptrace syscall & update strace

### PR DESCRIPTION
**Pull request by itamar8910**
This upgrades our tracing mechanism from `systrace` to `ptrace`.
The interface of `ptrace()` is based on [openbsd's](https://man.openbsd.org/ptrace.2) implementation.

In the future we could use `ptrace` for a debugger,
but right now it's only used by `strace`.

### CODE REVIEW

File: Kernel/Process.h [hunk 1]
```diff
@@ -585,7 +586,7 @@ inline void Process::for_each_child(Callback callback)
     pid_t my_pid = pid();
     for (auto* process = g_processes->head(); process;) {
         auto* next_process = process->next();
-        if (process->ppid() == my_pid) {
+        if (process->ppid() == my_pid || process->has_tracee_thread(m_pid)) {
```
[the remark this answers is not in the document]
Comment (itamar8910) [reply Kernel/Process.h]: Yeah I agree that the semantics are a bit off with this one.

The reason behind this change is to support `ptrace`ing a non-child process (which `strace -p` uses) - we need `waitpid` to also consider tracees (in addition to child processes).

`Proces::do_waitid` [invokes a `WaitBlocker`](https://github.com/SerenityOS/serenity/blob/f1a074a262a9815a1063b17d20df1d1a15308cef/Kernel/Process.cpp#L2386),
which [iterates the process' children](https://github.com/SerenityOS/serenity/blob/f1a074a262a9815a1063b17d20df1d1a15308cef/Kernel/Scheduler.cpp#L259)  and checks their states.

When we `ptrace` a process, we also want the `WaitBlocker` to unblock
if the tracee's state has changed.

Perhaps a better way of doing it is adding a `Process:for_each_tracee`?

[the remark this answers is not in the document]
Comment (awesomekling) [reply Kernel/Process.h]: Yes, that would be better. Alternatively `Process::for_each_child_and_tracee` if it makes more sense to make a single pass over the process table.

File: Kernel/ThreadTracer.h [hunk 2]
```diff
@@ -26,46 +26,44 @@
...
+    // signals that are sent from the tracer to the tracee
+    // TODO: currently it's always 0,
+    //       implement settings signals in here with PT_CONTINUE
+    u32 m_pending_signals {0};
 
-    pid_t m_pid;
-    bool m_dead { false };
-    CircularQueue<CallData, 200> m_calls;
+    bool m_trace_syscalls {false};
+    OwnPtr<PtraceRegisters> m_regs;
```
[the remark this answers is not in the document]
Comment (itamar8910) [reply Kernel/ThreadTracer.h]: We do not always hold a valid register set for the tracee - I used
`OwnPtr` to express this (a `null` value means no available registers).
Is there a better alternative in `AK`?

[the remark this answers is not in the document]
Comment (awesomekling) [reply Kernel/ThreadTracer.h]: Yes, `Optional<T>` :)

**Comment by awesomekling** [comment]
Okay let's merge it :) Thanks for working on this!

**Pull request merged**